Min max added to reduction ops

ShisuiUzumaki opened the pull request:
- [x] closes #50937 
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

ShisuiUzumaki committed:
assertion error fix added in pandas/core/indexes/base.py

ShisuiUzumaki committed:
Merge remote-tracking branch 'origin/main'

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Added `min` and `max` as reduction operators

ShisuiUzumaki commented:
Anybody who can review this: cc @myles @jd @troels @lebigot @baevpetr

mroeschke reviewed (changes requested):
Needs tests and a whatsnew note in `v2.0.0.rst`

ShisuiUzumaki commented:
> Needs tests and a whatsnew note in `v2.0.0.rst`

Could you give some pointers on how can I add these tests and whatsnew

mroeschke commented:
A unit test that checks the behavior of the related issue be added to `pandas/tests/frame/test_query_eval.py`.

An entry should be added to `doc/source/whatsnew/v2.0.0.rst` describing what was fixed

ShisuiUzumaki commented:
> A unit test that checks the behavior of the related issue be added to `pandas/tests/frame/test_query_eval.py`.
> 
> An entry should be added to `doc/source/whatsnew/v2.0.0.rst` describing what was fixed

Would the following test be okay?

```
class TestDataFrameQueryNumExprPython(TestDataFrameQueryNumExprPandas):

    ...

    def test_query_numexpr_with_min_and_max_columns(self):
        df = DataFrame({"min": [1, 2, 3], "max": [4, 5, 6]})
        regex_to_match = r"Variables in expression \"\(min\) == \(1\)\" overlap with builtins: \('min'\)"
        with pytest.raises(NumExprClobberingError, match=regex_to_match):
            df.query("min == 1")

        regex_to_match = r"Variables in expression \"\(max\) == \(1\)\" overlap with builtins: \('max'\)"
        with pytest.raises(NumExprClobberingError, match=regex_to_match):
            df.query("max == 1")

```

mroeschke commented:
Yes but the tests can just be stand alone functions

ShisuiUzumaki commented:
> An entry should be added to `doc/source/whatsnew/v2.0.0.rst` describing what was fixed

Also, could you give me a bit more hints on what format I should follow to add changes descriptions in 
` doc/source/whatsnew/v2.0.0.rst `. Like in what section, and what should be the phrasing format, etc.

mroeschke commented:
This will be easier to comment on if you commit the changes you suggested

ShisuiUzumaki commented:
> This will be easier to comment on if you commit the changes you suggested

Sure!

ShisuiUzumaki committed:
Added `test` and `whatsnew`

- Added `test_query_numexpr_with_min_and_max_columns`  in `pandas/frame/tests/test_query_eval.py`
- Bug fix entry added in `pandas/doc/source/whatsnew/v2.0.0.rst` for issue `#50937`

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

ShisuiUzumaki committed:
Merge branch 'main' of https://github.com/ShisuiUzumaki/pandas

mroeschke commented on doc/source/whatsnew/v2.0.0.rst:
```diff
@@ -1016,6 +1016,7 @@ Numeric
 - Bug in DataFrame reduction methods (e.g. :meth:`DataFrame.sum`) with object dtype, ``axis=1`` and ``numeric_only=False`` would not be coerced to float (:issue:`49551`)
 - Bug in :meth:`DataFrame.sem` and :meth:`Series.sem` where an erroneous ``TypeError`` would always raise when using data backed by an :class:`ArrowDtype` (:issue:`49759`)
 - Bug in :meth:`Series.__add__` casting to object for list and masked :class:`Series` (:issue:`22962`)
+- Bug in :meth:`query` with ``engine="numexpr"`` and column names are ``min`` or ``max``, does not work correctly (:issue:`50937`)
```
```suggestion
- Bug in :meth:`query` with ``engine="numexpr"`` and column names are ``min`` or ``max`` would raise a ``TypeError`` (:issue:`50937`)
```

ShisuiUzumaki replied:
> 

Why did we replace it with a `TypeError `rather than the error it raises

mroeschke commented:
```
************* Module pandas.tests.frame.test_query_eval
pandas/tests/frame/test_query_eval.py:540:8: W0404: Reimport 'NumExprClobberingError' (imported line 6) (reimported)
```

ShisuiUzumaki committed:
Update doc/source/whatsnew/v2.0.0.rst

Co-authored-by: Matthew Roeschke <10647082+mroeschke@users.noreply.github.com>

ShisuiUzumaki committed:
reimport error fixed

mroeschke commented:
Thanks @ShisuiUzumaki
